tests/quick.rs

```diff
@@ -0,0 +1,200 @@
+extern crate evmap;
+
+use evmap::{ReadHandle, WriteHandle};
+
+extern crate quickcheck;
+#[macro_use(quickcheck)]
+extern crate quickcheck_macros;
+
+use quickcheck::Arbitrary;
+use quickcheck::Gen;
+
+use rand::Rng;
+use std::cmp::{min, Ord};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::{BuildHasher, Hash};
+use std::ops::Deref;
+
+fn set<'a, T: 'a, I>(iter: I) -> HashSet<T>
+where
+    I: IntoIterator<Item = &'a T>,
+    T: Copy + Hash + Eq,
+{
+    iter.into_iter().cloned().collect()
+}
+
+#[quickcheck]
+fn contains(insert: Vec<u32>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    insert.iter().all(|&key| r.get(&key).is_some())
+}
+
+#[quickcheck]
+fn contains_not(insert: Vec<u8>, not: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    let nots = &set(&not) - &set(&insert);
+    nots.iter().all(|&key| r.get(&key).is_none())
+}
+
+#[quickcheck]
+fn insert_empty(insert: Vec<u8>, remove: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+    for &key in &remove {
+        w.empty(key);
+    }
+    w.refresh();
+    let elements = &set(&insert) - &set(&remove);
+    r.len() == elements.len()
+        && r.map_into::<_, Vec<()>, _>(|_, _| ()).len() == elements.len()
+        && elements.iter().all(|k| r.get(k).is_some())
+}
+
+use Op::*;
+#[derive(Copy, Clone, Debug)]
+enum Op<K, V> {
+    Add(K, V),
+    Remove(K),
+}
+
+impl<K, V> Arbitrary for Op<K, V>
+where
+    K: Arbitrary,
+    V: Arbitrary,
+{
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        match g.gen::<u32>() % 4 {
+            0 => Add(K::arbitrary(g), V::arbitrary(g)),
+            _ => Remove(K::arbitrary(g)),
+        }
+    }
+}
+
+fn do_ops<K, V, S>(ops: &[Op<K, V>], a: &mut WriteHandle<K, V, (), S>, b: &mut HashMap<K, Vec<V>>)
+where
+    K: Hash + Eq + Clone,
+    V: Clone + evmap::ShallowCopy + Eq + Hash,
+    S: BuildHasher + Clone,
+{
+    for op in ops {
+        match *op {
+            Add(ref k, ref v) => {
+                a.insert(k.clone(), v.clone());
+                b.entry(k.clone()).or_insert_with(Vec::new).push(v.clone());
+            }
+            Remove(ref k) => {
+                a.empty(k.clone());
+                b.remove(k);
+            }
+        }
+    }
+}
+
+fn assert_maps_equivalent<K, V, S>(a: &ReadHandle<K, V, (), S>, b: &HashMap<K, Vec<V>>) -> bool
+where
+    K: Clone + Hash + Eq + Debug,
+    V: Hash + Eq + Debug + Ord + Copy,
```

Comment:
This `Copy` and `Ord` are also laziness so I can easily sort and compare the values for each key between the maps. Am open to anything else!